src/visualizers/Chaos.ts

```diff
@@ -1,429 +1,804 @@
+        const endIndex = this.firstIndex + BigInt(targetLength)
+
+        let currElement = 0n
+        for (let i = startIndex; i < endIndex; i++) {
+            // try to get the new element and store its data
+            try {
+                currElement = this.seq.getElement(i)
+            } catch (e) {
+                this.pathFailure = true
+                if (e instanceof CachingError) {
+                    return // need to wait for more elements
+                } else {
+                    // don't know what to do with this
+                    throw e
+                }
+            }
+
+            // variables you can use in walker formula
+            // excludes C,c,w,x,y
+            const inputWalker = {
+                n: Number(i),
+                a: Number(currElement),
+                s: Number(i - this.seq.first + 1n),
+                m: Number(this.seq.first),
+                M: Number(this.seq.last),
+                f: currentFrames,
+                p: this.corners,
+                W: this.walkers,
+                A: (n: number | bigint) =>
+                    Number(this.seq.getElement(BigInt(n))),
+            }
+
+            // which walker do we move
+            let currWalker = 0
+            // gives bad data on bigints
+            currWalker = math.floor(
+                math.safeNumber(
+                    this.walkerFormula.computeWithStatus(
+                        this.statusOf.walkerFormula,
+                        inputWalker
+                    )
+                ) ?? -1 // invalid
             )
+            if (this.statusOf.walkerFormula.invalid()) return
+            if (currWalker < 0 || currWalker >= this.walkers) {
+                this.statusOf.walkerFormula.warnings.length = 0
+                this.statusOf.walkerFormula.addWarning(
+                    'some walkerFormula values not in walker range'
+                )
+                currWalker = -1 // invalid
+            }
+            if (currWalker == -1) {
+                continue
+            }
+
+            // look up last position of this walker
+            const currlen = this.dots[currWalker].length - 1
+            const position = this.dots[currWalker][currlen].copy()
+            const lastCorner = this.dotsCorners[currWalker][currlen]
+
+            // variables you can use in corner formula (excludes C)
+            const inputCorner = {
+                n: Number(i),
+                a: Number(currElement),
+                s: Number(i - this.seq.first + 1n),
+                m: Number(this.seq.first),
+                M: Number(this.seq.last),
+                f: currentFrames,
+                c: lastCorner,
+                p: this.corners,
+                w: currWalker,
+                W: this.walkers,
+                x: position.x,
+                y: position.y,
+                A: (n: number | bigint) =>
+                    Number(this.seq.getElement(BigInt(n))),
+            }
```

Comment:
Please modify the input object computations to re-use common values, rather than recompute them. That change will greatly aid maintenance of the code when we need to change the code used to compute one of the input variables, for example.

Reply:
I'm not sure I understand this comment.  Can you elaborate?  Or maybe it's just where you marked it that I'm finding it confusing.  Do you mean recomputing Number(i - this.seq.first + 1n)?  (This is marked next to the definition of A, which i don't see a way to improve.)

Reply:
I just mean you compute three different objects for inputs to formulas, that all share most of their property names and values, so the shared parts should only be computed once (for efficiency) and written once (for readability and maintainability). In other words, the three highly overlapping lists are not DRY.

Reply:
Now each input object incorporates the previous one and then adds the new values.